Set fonts via Preferences

boutinb opened the pull request:
Fixes jasp-stats/INTERNAL-jasp#1019

boutinb committed:
Set fonts via Preferences

JorisGoosen commented on JASP-Desktop/components/JASP/Theme/Theme.qml:
```diff
@@ -33,22 +33,22 @@ JaspTheme
 	fontLabel.bold:										true
 	fontLabel.pixelSize:								Math.round(18 * uiScale)
-	fontLabel.family:									jaspFont + (MACOS ? " Bold" : "")
+	fontLabel.family:									jaspFont
```
Did you make sure the font s are still bold/demibold etc when they ought to be?

boutinb replied:
I see no difference on my Mac if I remove this. It seems anyway quite redundant with the bold property.
And it must be removed because most of the fonts do not have a 'Bold' version: then it falls back to the default font.

JorisGoosen replied:
I think you would see the difference if you would specify the fonts as QFont instead of as QString, but if it works it works so whatever.

JorisGoosen commented on JASP-Desktop/components/JASP/Widgets/FileMenu/PrefsResults.qml:
```diff
@@ -130,12 +130,41 @@ ScrollView
+				id:					resultFont
+				width:				300 * jaspTheme.uiScale
+				height:				150 * jaspTheme.uiScale
+				title:				qsTr("List of Font")
```
"List of fonts"

JorisGoosen replied:
Also, can we just do this through a userselectable list of systemfonts instead of them having to type?

JorisGoosen commented on JASP-Desktop/components/JASP/Widgets/FileMenu/PrefsResults.qml:
```diff
@@ -130,12 +130,41 @@ ScrollView
+				width:				300 * jaspTheme.uiScale
+				height:				150 * jaspTheme.uiScale
+				title:				qsTr("List of Font")
+				toolTip:			qsTr("Font family used by the results")
```
If it is just for the results then why is it changing QML?

JorisGoosen commented on JASP-Desktop/components/JASP/Widgets/FileMenu/PrefsUI.qml:
```diff
@@ -69,12 +69,42 @@ ScrollView
+				text:				preferencesModel.interfaceFont.replace(/,/g, "\n")
+				onApplyRequest:
+				{
+					preferencesModel.interfaceFont = text.split("\n")
```
Code duplication...
Could you not do that? 

Maybe move the function to C++ and use it there?
Although if you change to a combobox this will probably not be necessary anymore.

JorisGoosen commented on JASP-Desktop/gui/preferencesmodel.h:
```diff
@@ -38,7 +38,8 @@ class PreferencesModel : public QObject
 	Q_PROPERTY(QString		cranRepoURL				READ cranRepoURL				WRITE setCranRepoURL				NOTIFY cranRepoURLChanged				)
 	Q_PROPERTY(int			plotPPI					READ plotPPI														NOTIFY plotPPIPropChanged				)
-	Q_PROPERTY(QFont		defaultFont				READ defaultFont				WRITE setDefaultFont				NOTIFY defaultFontChanged				)
+	Q_PROPERTY(QString		interfaceFont			READ interfaceFont				WRITE setInterfaceFont				NOTIFY interfaceFontChanged				)
```
Also a problem, by changing QFont to QString we will no longer be able to use non-system fonts.
Like the open source sansserif font included with JASP.
Which it would be great to use!

Because it contains all mathematical symbols for sure.

boutinb replied:
But we do not load FreeSans yet. And I think if we add it in the application, we can call it by its name (we will have to anyway, by selecting it in the Dropdown).

JorisGoosen replied:
Ooh yeah, ok. 

But I still think QFont is the way to go.

JorisGoosen commented on JASP-Desktop/qquick/jasptheme.h:
```diff
@@ -563,6 +563,7 @@ public slots:
 	void setIconPath(QString iconPath);
 	void setThemeName(QString themeName);
 	void setFontConsole(QFont fontConsole);
+	void setDefaultFont(QString font);
```
Why the QString and not the QFont?

JorisGoosen commented on Resources/Help/preferences/prefsui.md:
```diff
@@ -32,3 +32,6 @@ If this is enabled JASP will run in software rendering mode, which means your in
 If you disable this option we use Qt file dialogs and they might not crash. Most users will not need this option disabled though.
+
+### Font
+You can specify here which font will be used by the interface. Default is `SansSerif`
```
"Here you can specify which font will be used by the interface. Default is `SansSerif`"

JorisGoosen replied:
Should this also be added to the results thingy?

JorisGoosen commented:
I think having a combobox would greatly improve this functionality, in what application do users have to guess the name of the font they want to use?
And I think using a QFont makes more sense then QString, especially since this is one of the few datatypes incuded with QML.

JorisGoosen commented:
And I didn't even test it yet ;)

boutinb commented:
I see that Qt can retrieve the system fonts https://doc.qt.io/qt-5/qfontdatabase.html. I can try to use that.
But what I understand from fonts is that sometimes, the browser can decide to use different fonts for the same string if part of the string cannot be displayed with the first specified font, but with the second one. That's why there is a list of fonts. So one ComboBox is not enough. I could make a list of ComboBox...

JorisGoosen commented:
> I see that Qt can retrieve the system fonts https://doc.qt.io/qt-5/qfontdatabase.html. I can try to use that.

Good!

> But what I understand from fonts is that sometimes, the browser can decide to use different fonts for the same string if part of the string cannot be displayed with the first specified font, but with the second one. That's why there is a list of fonts. So one ComboBox is not enough. I could make a list of ComboBox...

I have never heard of a browser using different fonts for different parts of a string. Those lists are there for when a system doesn't have the first font, then it tries the second one. But if the user selects from systemfonts how can they not have that font?

Of course, if Im wrong about this I would like to hear it but I don't think so.

boutinb commented:
https://stackoverflow.com/questions/884177/how-can-i-determine-what-font-a-browser-is-actually-using-to-render-some-text
According to this discussion, the browser may choose another font if the current one does not support some characters. It's still a bit strange: if it was true, why do we get sometimes empty squares for some non-latin characters...
I have another issue: for the WebEngine we have now a list of fonts (Lucida Grande,Helvetica,Arial,sans-serif,Helvetica Neue,freesans,Segoe UI"). I could set it to Lucinda Grande per default, but if the computer does not have this font, it will get the default font, which is maybe different than the Helvetica (what he gets now if the PC has this font). Maybe small issue...
In QML it is now per default SansSerif, which does not exist on my Mac, and it gives apparently the default font ".SF NS Text" (which is per chance a Sans Serif font). I think I have to add a 'default system font' option.

JorisGoosen commented:
> https://stackoverflow.com/questions/884177/how-can-i-determine-what-font-a-browser-is-actually-using-to-render-some-text
> According to this discussion, the browser may choose another font if the current one does not support some characters. It's still a bit strange: if it was true, why do we get sometimes empty squares for some non-latin characters...

Because Arial on the user's system might be different because who knows what Microsoft does to support other locales?

> I have another issue: for the WebEngine we have now a list of fonts (Lucida Grande,Helvetica,Arial,sans-serif,Helvetica Neue,freesans,Segoe UI"). I could set it to Lucinda Grande per default, but if the computer does not have this font, it will get the default font, which is maybe different than the Helvetica (what he gets now if the PC has this font). Maybe small issue...
> In QML it is now per default SansSerif, which does not exist on my Mac, and it gives apparently the default font ".SF NS Text" (which is per chance a Sans Serif font). I think I have to add a 'default system font' option.

SansSerif is a generic font family which allows someone to not have to specify exactly what font to use.

I think having an option called "default" for the results is fine, it would just not use anything special. And to get back to it after the user changed something you could just reload the css.

boutinb committed:
Use a dropdown with all possible fonts

. Add also the SansFree font to the list
. Set the font also in the Help webengine

JorisGoosen commented:
Nice!
Works pretty awesome :)

And it is hilarious to see JASP change as I move through the different fonts for the interface.

I did find a small bug still with the "default" option for the results. It says next to it that the "default" is "Deja vu sans" which Im guessing is the system sansserif I have. But if I turn the checkbox on the results turn to a serif font...

A la:
![Screenshot_2020-08-19_173819](https://user-images.githubusercontent.com/29062713/90657515-f5561600-e242-11ea-89e8-6a15ea2c12b1.png)

Then to the actual selected font:
![Screenshot_2020-08-19_173823](https://user-images.githubusercontent.com/29062713/90657526-f6874300-e242-11ea-9d78-b86dca8f422f.png)

Then I tried restarting with "default" on but it shows it with the selected font instead:
![Screenshot_2020-08-19_173906](https://user-images.githubusercontent.com/29062713/90657535-f71fd980-e242-11ea-9fce-abe61f9395a8.png)

So there is definitely something still wrong there.

boutinb commented:
I think the problem is that the DejaVu Sans is not recognized by The WebEngine. Could you start the Chrome inspector and ask what the value of the actual font (in computed tab, at the bottom), and whether the body gets the DejaVu Sans font?

JorisGoosen commented:
I think there are two problems:

1. When I start JASP and load a file while "default font" is checked for results it still uses the dropdown font. So that is bad
2. When I turn default off and on again it switches to a different font than selected in dropdown. But it is: "font-family: SansSerif;"

What I would suggest for turning on default font is: simply reload the "theme css" and then *not set any font*. Is that an option?

JorisGoosen commented on JASP-Desktop/gui/preferencesmodel.cpp:
```diff
@@ -33,6 +34,13 @@ PreferencesModel::PreferencesModel(QObject *parent) :
 
 	connect(LanguageModel::lang(),	&LanguageModel::currentIndexChanged,			this, &PreferencesModel::languageCodeChanged);
+
+	connect(this,					&PreferencesModel::useDefaultInterfaceFontChanged,	[&](){	emit realInterfaceFontChanged(realInterfaceFont()); } );
```
Are you aware you can connect signals to signals? Would be prettier than this.
Also, I don't think it is necessary to send the font with it?

JorisGoosen commented on JASP-Desktop/gui/preferencesmodel.cpp:
```diff
@@ -107,7 +115,41 @@ GET_PREF_FUNC_STR(	currentThemeName,			Settings::THEME_NAME								)
+		// The default is a list of fonts.
+		// Select the first one which is available.
+		for (const QString& font : result.split(","))
+		{
```
Why the braces?

JorisGoosen commented on JASP-Desktop/results/resultsjsinterface.cpp:
```diff
@@ -73,6 +73,7 @@ void ResultsJsInterface::setResultsLoaded(bool resultsLoaded)
 		emit runJavaScript("window.setAppVersion('" + version + "')");
 
 		setGlobalJsValues();
+		setFontFamily(Settings::value(Settings::RESULT_FONT).toString());
```
Just check here if "default is on" and if so don't set the result font especially.

JorisGoosen commented:
I think it would work better to set the results font in the way Ive described and then  you can also remove the "Deja Vu Sans" or whatever font next to it. Because it might not be the right font anyway. 
Just show it as "default" instead.

JorisGoosen commented:
And a little bit of an extension: maybe we could also make the font for the console/code-windows changeable?

JorisGoosen commented:
(After the code actually works I will do an in-depth code review btw ;) )

boutinb committed:
Add R console fonts, etc...

. Fix default font when starting JASP
. Set default font setting for result at one place
. Add R Console fonts
. UI to set the 3 fonts is now grouped onder Interface preferences

JorisGoosen commented:
I think once you have the FiraCode commited this can be merged!

boutinb committed:
Add R Console, JAGS, Lavaan font in preference
